src/coreComponents/physicsSolvers/fluidFlow/DissipationCompositionalMultiphaseFVMKernels.hpp

```diff
@@ -0,0 +1,469 @@
+  {}
+
+  struct StackVariables : public Base::StackVariables
+  {
+public:
+
+    GEOS_HOST_DEVICE
+    StackVariables( localIndex const size, localIndex numElems )
+      : Base::StackVariables( size, numElems )
+    {}
+
+    using Base::StackVariables::stencilSize;
+    using Base::StackVariables::numConnectedElems;
+    using Base::StackVariables::transmissibility;
+    using Base::StackVariables::dTrans_dPres;
+    using Base::StackVariables::dofColIndices;
+    using Base::StackVariables::localFlux;
+    using Base::StackVariables::localFluxJacobian;
+
+  };
+
+  /**
+   * @brief Compute the local flux contributions to the residual and Jacobian, including dissipation
+   * @param[in] iconn the connection index
+   * @param[inout] stack the stack variables
+   */
+  GEOS_HOST_DEVICE
+  void computeFlux( localIndex const iconn,
+                    StackVariables & stack ) const
+  {
+
+    m_stencilWrapper.computeWeights( iconn,
+                                     m_permeability,
+                                     m_dPerm_dPres,
+                                     stack.transmissibility,
+                                     stack.dTrans_dPres );
+
+    // ***********************************************
+    // First, we call the base computeFlux to compute:
+    //  1) compFlux and its derivatives,
+    //
+    // We use the lambda below (called **inside** the phase loop of the base computeFlux) to compute dissipation terms
+    Base::computeFlux( iconn, stack, [&] ( integer const ip,
+                                           localIndex const (&k)[2],
+                                           localIndex const (&seri)[2],
+                                           localIndex const (&sesri)[2],
+                                           localIndex const (&sei)[2],
+                                           localIndex const connectionIndex,
+                                           localIndex const k_up,
+                                           localIndex const er_up,
+                                           localIndex const esr_up,
+                                           localIndex const ei_up,
+                                           real64 const & potGrad,
+                                           real64 const & phaseFlux,
+                                           real64 const (&dPhaseFlux_dP)[2],
+                                           real64 const (&dPhaseFlux_dC)[2][numComp] )
+    {
+      GEOS_UNUSED_VAR( k_up, potGrad, phaseFlux, dPhaseFlux_dP, dPhaseFlux_dC, er_up, esr_up, ei_up );
+
+      /// Storing dissipation flux and its derivatives locally
+      real64 dissFlux[numComp]{};
+      real64 dDissFlux_dP[numFluxSupportPoints][numComp]{};
+      real64 dDissFlux_dC[numFluxSupportPoints][numComp][numComp]{};
+      real64 fluxPointCoef[numFluxSupportPoints] = {1.0, -1.0}; // for gradients
+
+      real64 viscosityMult[3] = {1.0, 1.0, 1.0}; // for viscosity
```

Comment:
Is local multipliers `viscosityMult` likely to be altered to another value than `{1.,1.,1.}` ?

Reply:
@jafranc The idea behind these viscosity multipliers was to include the effect of the phase viscosities  (different dissipation flux for different phases). However, the viscosities of the phases from the previous timestep (time n) are not stored in GEOS (at least to my knowledge), so I couldn't add them (and so kept them at 1.0). I've tried different values for these multipliers (by hard coding), but did not see an order 1 impact on the nonlinear convergence. And then, of course, more testing and sensitivity is to be done to have an adaptive tuning for them.